base64ct/src/base64.rs

```diff
@@ -0,0 +1,229 @@
+//! # Base64
+//!
+//! Base64 encoding as described in [RFC 4648, section 4].
+//!
+//! [RFC 4648, section 4]: https://tools.ietf.org/html/rfc4648#section-4
+
+use crate::errors::{InvalidEncodingError, InvalidLengthError};
+use core::str;
+
+#[cfg(feature = "alloc")]
+use alloc::{string::String, vec::Vec};
+
+/// Encode
+pub fn encode(src: &[u8], dst: &mut [u8]) -> Result<usize, InvalidLengthError> {
+    if encoded_len(src) > dst.len() {
+        return Err(InvalidLengthError);
+    }
+
+    let mut src_offset: usize = 0;
+    let mut dst_offset: usize = 0;
+    let mut src_length: usize = src.len();
+
+    while src_length >= 3 {
+        encode_3bytes(
+            &src[src_offset..(src_offset + 3)],
+            &mut dst[dst_offset..(dst_offset + 4)],
+        );
+
+        src_offset += 3;
+        dst_offset += 4;
+        src_length -= 3;
+    }
+
+    if src_length > 0 {
+        let mut tmp = [0u8; 3];
+        tmp[..src_length].copy_from_slice(&src[src_offset..(src_offset + src_length)]);
+        encode_3bytes(&tmp, &mut dst[dst_offset..]);
+        // tmp.zeroize();
+
+        dst[dst_offset + 3] = b'=';
+
+        if src_length == 1 {
+            dst[dst_offset + 2] = b'=';
+        }
+
+        dst_offset += 4;
+    }
+
+    Ok(dst_offset)
+}
+
+/// Encoded len
+pub fn encoded_len(bytes: &[u8]) -> usize {
+    (((bytes.len() * 4) / 3) + 3) & !3
+}
+
+/// Encode the input byte slice as a Base64-encoded [`String`].
+#[cfg(feature = "alloc")]
+#[cfg_attr(docsrs, doc(cfg(feature = "alloc")))]
+pub fn encode_string(input: &[u8]) -> String {
+    let elen = encoded_len(input);
+    let mut dst = vec![0u8; elen];
+    let res = encode(input, &mut dst);
+    debug_assert_eq!(elen, res.unwrap());
+    debug_assert!(str::from_utf8(&dst).is_ok());
+    // SAFETY: `dst` is fully written and contains only valid one-byte UTF-8 chars
+    unsafe { String::from_utf8_unchecked(dst) }
+}
+
+/// Decode
+pub fn decode(src: impl AsRef<[u8]>, dst: &mut [u8]) -> Result<usize, InvalidEncodingError> {
+    let src = src.as_ref();
+
+    // TODO(tarcieri): constant-time whitespace tolerance
+    if !src.is_empty() && char::from(src[src.len() - 1]).is_whitespace() {
```

Comment:
Are you sure you want to be so generic instead of simply checking for `b' '`?

Reply:
It should definitely handle newlines as well. They're probably the most common issue here, actually.

All that said, unless there's a separate error variant for whitespace versus other characters, having an eager whitespace check probably provides little value. So we should either add another error variant, or just remove this check entirely.